src/blockdata/witness.rs

```diff
@@ -0,0 +1,328 @@
+//! Witness
+//!
+//! This module contains the [`Witness`] struct and relative methods to operate on it
+//!
+
+use consensus::encode::{Error, MAX_VEC_SIZE};
+use consensus::{Decodable, Encodable, WriteExt};
+use io::{self, Read, Write};
+use prelude::*;
+use VarInt;
+
+#[cfg(feature = "serde")] use serde;
+
+/// The Witness is the data used to unlock bitcoins since the [segwit upgrade](https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki)
+///
+/// Can be logically seen as an array of byte-arrays `Vec<Vec<u8>>` and indeed you can convert from
+/// it and collect the iteration to convert into it.
+/// For serialization and deserialization performance it is stored internally as a single `Vec`,
+/// saving some allocations
+///
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+pub struct Witness {
+    /// contains the witness Vec<Vec<u8>> serialization without the initial varint indicating the
+    /// number of elements (which is stored in len)
+    content: Vec<u8>,
+
+    /// Number of elements in the witness.
+    /// It is stored separately (instead of as VarInt in the initial part of content) so that method
+    /// like [`Witness::push`] doesn't have case requiring to shift the entire array
+    witness_elements: u64,
+}
+
+/// Support structure to allow efficient and convenient iteration over the Witness elements
+pub struct WitnessIterator<'a> {
+    witness: &'a Witness,
+    cursor: usize,
+}
+
+impl From<Vec<Vec<u8>>> for Witness {
```

Comment:
In c342adbb1de1adc3dfc52f1b1cf8116f1d37b049:

I think the `From` trait is inappropriate here because you're doing a new allocation and the conversion is nontrivial. Can you pull this into an independent `from_vec` or something method? (Also, could you add a `to_vec` method which is ust a convenience wrapper over iterate-and-collect?)

Relatedly, in your serde `Deserialize` impl, I'm sure we can do something more efficient than decoding a `Vec<Vec<u8>>` and then converting -- though for this PR that's fine, we can do optimizations later.

Reply:
I added 8d23404a268eba662ac97a3292cf7247ce974c5d
If the commit is what you expected I guess I'll squash the last 4 commits in the first

Reply:
Not going to squash since it received the ACK it looks it's fine like it is